examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1091 @@
+                          MeshWorker::assemble_own_cells);
+  }
+
+
+  template <int dim>
+  std::unique_ptr<MGSmoother<Vector<double>>>
+  AdvectionProblem<dim>::create_smoother()
+  {
+    if (settings.smoother_type == "sor")
+      {
+        typedef PreconditionSOR<SparseMatrix<double>> Smoother;
+
+        auto smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices,
+                             Smoother::AdditionalData(fe.degree == 1 ? 1.0 :
+                                                                       0.62));
+        smoother->set_steps(2);
+        return smoother;
+      }
+    else if (settings.smoother_type == "jacobi")
+      {
+        typedef PreconditionJacobi<SparseMatrix<double>> Smoother;
+        auto                                             smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices,
+                             Smoother::AdditionalData(fe.degree == 1 ? 0.6667 :
+                                                                       0.47));
+        smoother->set_steps(4);
+        return smoother;
+      }
+    else if (settings.smoother_type == "block sor")
+      {
+        typedef RelaxationBlockSOR<SparseMatrix<double>, double, Vector<double>>
+          Smoother;
+
+        static MGLevelObject<typename Smoother::AdditionalData> smoother_data;
+        smoother_data.resize(0, triangulation.n_levels() - 1);
+
+        for (unsigned int level = 0; level < triangulation.n_levels(); ++level)
+          {
+            DoFTools::make_cell_patches(smoother_data[level].block_list,
+                                        dof_handler,
+                                        level);
+
+            smoother_data[level].relaxation = 1.0;
+            smoother_data[level].inversion = PreconditionBlockBase<double>::svd;
+
+            std::vector<unsigned int> ordered_indices;
+            if (settings.dof_renum == "downstream")
+              ordered_indices = create_downstream_order(dof_handler,
+                                                        advection_direction,
+                                                        level);
+            else if (settings.dof_renum == "upstream")
+              ordered_indices =
+                create_downstream_order(dof_handler,
+                                        -1.0 * advection_direction,
+                                        level);
+            else if (settings.dof_renum == "random")
+              ordered_indices = create_random_order(dof_handler, level);
+            else if (settings.dof_renum == "none")
```

Comment:
you could avoid these sequences of `if... else if...` if you parsed `dof_renum` and used an `enum` so that you can here use a `switch...case...case` construct.

As a side note, spell out `dof_renum(bering)`.